libs/core/synchronization/include/hpx/synchronization/detail/range_lock_impl.hpp

```diff
@@ -0,0 +1,106 @@
+#pragma once
+
+#include <hpx/execution_base/this_thread.hpp>
+
+#include <atomic>
+#include <boost/container/flat_map.hpp>
+#include <cstddef>
+#include <utility>
+#include <vector>
+
+namespace hpx::synchronization::detail {
+
+    template <typename Lock, template <typename> typename Guard>
+    class RangeLock
```

Comment:
Can we call this `range_mutex` instead? The name `lock` is normally used for the lock guards, not the actual 'locks'.